js/tools/gradientBrush.js

```diff
@@ -0,0 +1,50 @@
+window.tools.gradientBrush = {
+  name: 'gradientBrush',
+  icon: '/images/gradient-icon.png',
+  state: {
+    selected: false,
+    mousePressed: false,
+    prevMouse: { x: null, y: null },
+    baseColor: { g: 150, b: 150 }
+  },
+  events: {
+    mousedown: function (e, self) {
+      self.state.mousePressed = true
+    },
+    mouseup: function (e, self) {
+      self.state.mousePressed = false
+      self.state.prevMouse = { x: null, y: null }
+      self.state.baseColor = getRandomColor();
+    },
+    mousemove: function (e, self) {
+    // if self tool is selected AND the mouse is pressed
+      if (self.state.selected && self.state.mousePressed) {
+        const mouse = app.eventToMouse(e)
+        const px = self.state.prevMouse.x || mouse.x
+        const py = self.state.prevMouse.y || mouse.y
+
+        // draw a line
+        app.ctx.beginPath()
+        app.ctx.moveTo(mouse.x, mouse.y)
+        app.ctx.lineTo(px, py)
+        app.ctx.closePath()
+        app.ctx.stroke()
+        for (let i = 0; i < 255; i++)
+        {
+          app.ctx.beginPath()
+          app.ctx.moveTo(mouse.x + i, mouse.y)
+          app.ctx.lineTo(px + i, py)
+          app.ctx.closePath()
+          app.ctx.strokeStyle = `rgb(${i},${self.state.baseColor.g}, ${self.state.baseColor.b})`;
+          app.ctx.stroke()
+        }
+        // update prevMouse coordinates
+        self.state.prevMouse = { x: mouse.x, y: mouse.y }
+      }
+    }
+  }
+}
+
+function getRandomColor () {
+  return { g: Math.floor(Math.random() * 256), b: Math.floor(Math.random() * 256) }
+}
```

Comment:
by declaring a function outside the scope of ur tool object here, u're essentially introducing it to the entire project as a "global" variable, this is dangerous b/c if anyone else creates a `getRandomColor` variable (be it another function or anything else) it will conflict w/urs. what u should do instead is define it w/in the scope of ur tool, there are two ways to do this:

**the first way**

b/c ur random color function is only used inside our mouseup event, u could simply define it in there so that it's a local variable scoped to that function (it won't exist anywhere else in the code base)

```js
    mouseup: function (e, self) {
      function getRandomColor () {
        return { g: Math.floor(Math.random() * 256), b: Math.floor(Math.random() * 256) }
      }

      self.state.mousePressed = false
      self.state.prevMouse = { x: null, y: null }
      self.state.baseColor = getRandomColor()
    },
 ```
 
 **the second way**
 
 if in the future u want to create a function that u use w/in different events or w/in other functions in ur code base, then u can define it as a custom property in ur tool object like this:
 
 ```js
 window.tools.gradientBrush = {
  // name, icon, state etc go here...
  getRandomColor: function () {
    return { g: Math.floor(Math.random() * 256), b: Math.floor(Math.random() * 256) }
  },
  events: {
    mouseup: function (e, self) {
      self.state.mousePressed = false
      self.state.prevMouse = { x: null, y: null }
      self.state.baseColor = self.getRandomColor()
    },
    // ...other events go here
  }
}
 ```
 
 note that in this example i use the keyword `self` to access the `getRandomColor` function b/c it is now one of the tool's properties

Reply:
Done